### PR TITLE
feat!: volume agnostic material mapping

### PR DESCRIPTION
**Pull request by asalzburger**
This PR changes the material mapper logic and cleans up the code quite a bit.

While the algorithmic code rests identical (output unchanged), the `SurfaceMaterialMapper` and `VolumeMaterialMapper` have now entirely separate functionality:

  * the surface mapper does only surface mapping and already implements `TrackingGeometry` and `Detector` schema
  * the volume mapper does only volume mapping (currently not yet fully supporting the `Detector` schema)

It does a small renaming:
  * `MappingType` -> `MaterialMappingType`
In order to harmonise with:
  * `MaterialMappingState` the state base class
  * `MaterialMappingResult` the surface/volume maps produced

The Mappers are now designed to run in a chained setup, i.e. one mapper runs and returns a pair of `mapped` and `unmapped` tracks, the `unmapped` part is then processed by the subsequent mapper.

By doing so, the Mappers became `const correct` (except for the finalizeMaps) call, as modification of the input collection is not allowed anymore. 

Execution time / json map of vanilla ODD *before* change:
```
09:48:55    Sequencer      INFO      Processed 1000 events in 8.657600 s (wall clock)
09:48:55    Sequencer      INFO      Average time per event: 8.594614 ms/event
```

```
-rw-r--r--  1 salzburg  staff  238417544 Sep 22 09:48 material-map.json
```

Execution time / json map of vanilla ODD *after* change:
```
09:49:33    Sequencer      INFO      Processed 1000 events in 8.856570 s (wall clock)
09:49:33    Sequencer      INFO      Average time per event: 8.160427 ms/event
```

```
-rw-r--r--  1 salzburg  staff  238417544 Sep 22 09:49 material-map.json
```

It also includes a comprehensive set of new UnitTests for the `SurfacesMaterialMapper` which reflects in a sizeable increase of test coverage.

The parameter optimisation and the full validation of the `VolumeMaterialMapper` as well as the follow-up for the `Detector` model will come in separate PRs.

### CODE REVIEW

**Comment by asalzburger** [comment]
The example failures are in the material mapping, so probably not a fluke - I will check this early next week.

File: Tests/UnitTests/Core/Material/SurfaceMaterialMapperTests.cpp [hunk 1]
```diff
@@ -1,134 +1,218 @@
...
+  BOOST_CHECK(surfaceState != nullptr);
+  BOOST_CHECK(surfaceState->accumulatedMaterial.size() == 6u);
+
+  // Run the mapping
+  auto mTrack = createRecordedMaterialTrack(materialSteps);
+  auto [mapped, unmapped] = smMapper.mapMaterialTrack(*surfaceState, mTrack);
+  // Check that all material is mapped
+  BOOST_CHECK(mapped.second.materialInteractions.size() ==
+              mTrack.second.materialInteractions.size());
+  BOOST_CHECK(unmapped.second.materialInteractions.size() == 0u);
```
Comment (AJPfleger) [inline Tests/UnitTests/Core/Material/SurfaceMaterialMapperTests.cpp:91]:
```suggestion
  BOOST_REQUIRE(surfaceState != nullptr);
  BOOST_CHECK_EQUAL(surfaceState->accumulatedMaterial.size(), 6u);

  // Run the mapping
  auto mTrack = createRecordedMaterialTrack(materialSteps);
  auto [mapped, unmapped] = smMapper.mapMaterialTrack(*surfaceState, mTrack);
  // Check that all material is mapped
  BOOST_CHECK_EQUAL(mapped.second.materialInteractions.size(),
              mTrack.second.materialInteractions.size());
  BOOST_CHECK_EQUAL(unmapped.second.materialInteractions.size(), 0u);
```

File: Tests/UnitTests/Core/Material/SurfaceMaterialMapperTests.cpp [hunk 2]
```diff
@@ -1,134 +1,218 @@
...
+  auto [remapped, reunmapped] = smMapper.mapMaterialTrack(*restate, mapped);
+
+  BOOST_CHECK(mapped.second.materialInteractions.size() ==
+              remapped.second.materialInteractions.size());
+  CHECK_CLOSE_ABS(mapped.second.materialInX0, remapped.second.materialInX0,
+                  1e-10);
+  CHECK_CLOSE_ABS(mapped.second.materialInL0, remapped.second.materialInL0,
+                  1e-10);
+  BOOST_CHECK(unmapped.second.materialInteractions.size() ==
+              reunmapped.second.materialInteractions.size());
```
Comment (AJPfleger) [inline Tests/UnitTests/Core/Material/SurfaceMaterialMapperTests.cpp:110]:
```suggestion
  BOOST_CHECK_EQUAL(mapped.second.materialInteractions.size(),
              remapped.second.materialInteractions.size());
  CHECK_CLOSE_ABS(mapped.second.materialInX0, remapped.second.materialInX0,
                  1e-10);
  CHECK_CLOSE_ABS(mapped.second.materialInL0, remapped.second.materialInL0,
                  1e-10);
  BOOST_CHECK_EQUAL(unmapped.second.materialInteractions.size(),
              reunmapped.second.materialInteractions.size());
```

File: Tests/UnitTests/Core/Material/SurfaceMaterialMapperTests.cpp [hunk 3]
```diff
@@ -1,134 +1,218 @@
...
+  std::for_each(materialSteps.begin(), materialSteps.end(),
+                [&](const Acts::ActsScalar& p) {
+                  if (p < 0)
+                    negativeSide++;
+                });
+
+  // Check that all material is mapped
+  BOOST_CHECK(mapped.second.materialInteractions.size() == negativeSide);
+  BOOST_CHECK(unmapped.second.materialInteractions.size() ==
+              mTrack.second.materialInteractions.size() - negativeSide);
```
Comment (AJPfleger) [inline Tests/UnitTests/Core/Material/SurfaceMaterialMapperTests.cpp:152]:
```suggestion
  BOOST_CHECK_EQUAL(mapped.second.materialInteractions.size(), negativeSide);
  BOOST_CHECK_EQUAL(unmapped.second.materialInteractions.size(),
              mTrack.second.materialInteractions.size() - negativeSide);
```

File: Tests/UnitTests/Core/Material/SurfaceMaterialMapperTests.cpp [hunk 4]
```diff
@@ -1,134 +1,218 @@
...
+      Acts::getDefaultLogger("PlainSurfaceMaterialMapper",
+                             Acts::Logging::VERBOSE));
+
+  auto pstate = psmMapper.createState(tContext, tContext, *tGeometry);
+  auto mTrack = createRecordedMaterialTrack(materialSteps);
+  auto [mapped, unmapped] = psmMapper.mapMaterialTrack(*pstate, mTrack);
+
+  // Check that all material is mapped
+  BOOST_CHECK(mapped.second.materialInteractions.size() ==
+              mTrack.second.materialInteractions.size());
```
Comment (AJPfleger) [inline Tests/UnitTests/Core/Material/SurfaceMaterialMapperTests.cpp:186]:
```suggestion
  BOOST_CHECK_EQUAL(mapped.second.materialInteractions.size(),
              mTrack.second.materialInteractions.size());
```

File: Tests/UnitTests/Core/Material/SurfaceMaterialMapperTests.cpp [hunk 5]
```diff
@@ -1,134 +1,218 @@
...
+  std::size_t negativeSide = 0;
+  std::for_each(materialSteps.begin(), materialSteps.end(),
+                [&](const Acts::ActsScalar& p) {
+                  if (p < 0)
+                    negativeSide++;
+                });
+  // Check that all material is mapped
+  BOOST_CHECK(remapped.second.materialInteractions.size() == negativeSide);
+  BOOST_CHECK(reunmapped.second.materialInteractions.size() ==
+              mTrack.second.materialInteractions.size() - negativeSide);
```
Comment (AJPfleger) [inline Tests/UnitTests/Core/Material/SurfaceMaterialMapperTests.cpp:215]:
```suggestion
  BOOST_CHECK_EQUAL(remapped.second.materialInteractions.size(), negativeSide);
  BOOST_CHECK_EQUAL(reunmapped.second.materialInteractions.size(),
              mTrack.second.materialInteractions.size() - negativeSide);
```

**Review by AJPfleger — CHANGES_REQUESTED** [review]

**Pull request closed without merging**